### PR TITLE
Fix convert operation for NaNs

### PR DESCRIPTION
**Pull request by umangyadav**
Ref implementation of the `convert` op in MIGraphX was using clamping along with type-casting.  It doesn't work for NaNs.

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1758 Added a verify test with `double` type literal NaNs. 

MIGraphX converts double-typed literals to float. This convert operation gets const-folded at compile time using Ref Implementation and doesn't result in NaNs. That resulted in unit-test failures and was reported by QA.

For some reason, this error is only re-producible on certain machines. But I can verify using the following small test that `type-casting+clamping` doesn't work for NaNs. 

```cpp
#include <limits>
#include <iostream>
#include <cmath>


int main() {
    double d = std::numeric_limits<double>::quiet_NaN();
    float converted_f = std::min(std::numeric_limits<float>::max(), std::max(std::numeric_limits<float>::min(), float(d)));
    std::cout << "is nan: " << std::isnan(d) << "\n";
    std::cout << "is converted value a nan: " << std::isnan(converted_f) << "\n";
    std::cout << "is casted value a nan:  " << std::isnan(float(d)) << "\n";
    return 0;
}
```

Running that program print the following: 

```
is nan: 1
is converted value a nan: 0
is casted value a nan:  1
```

### CODE REVIEW

File: test/ref_ops_test.cpp [hunk 1]
```diff
@@ -1849,6 +1849,40 @@ TEST_CASE(cosh_dyn_test)
     EXPECT(migraphx::verify_range(results_vector, gold));
 }
 
+TEST_CASE(convert_nan_upcast_test)
```
Comment (TedThemistokleous) [inline test/ref_ops_test.cpp:1852]: Add a test case for half as well and even a double->float->half to ensure things stay preserved.

**Review by TedThemistokleous — COMMENTED** [review]
Add another two  test cases going between the three conversions up and down

**Comment by TedThemistokleous** [comment]
@umangyadav what machines was this failing on? What about the other direction of clamping using:

```
float converted_f_upper_clamp = std::max(std::numeric_limits<float>::min(), std::min(std::numeric_limits<float>::max(), float(d)));
```

File: test/ref_ops_test.cpp (hunk 1 above)
Comment (umangyadav) [reply test/ref_ops_test.cpp:1852]: I don't see the need to add a Double convert test (a) it is very unlikely to have triple converts inside a model. (b) Double converts are just two convert operations in tandem. I've individual convert tests, if they work, then double converts would work for sure. 

For the GPU target, we run simplify_algebra/reshapes. Which would convert such double converts to single convert. It doesn't happen for "ref" though.  

I added extra two tests.

**Comment by umangyadav** [comment]
> @umangyadav what machines was this failing on? What about the other direction of clamping using:
> 
> ```
> float converted_f_upper_clamp = std::max(std::numeric_limits<float>::min(), std::min(std::numeric_limits<float>::max(), float(d)));
> ```

It doesn't work either.

**Review by TedThemistokleous — APPROVED** [review]

**Review by kahmed10 — APPROVED** [review]

**Pull request merged**